### PR TITLE
Switch to OpenPGP v5

### LINKED ISSUE #3324: switch browser extension to OpenPGP v5
Not ready yet, but hopefully in a few months it will be

### LINKED ISSUE #4904: could key be sent to background page binary / unarmored?
Would it be faster to send the key in binary representation (unarmored openpgp key)? Armoring/de-armoring is oddly expensive, about 10ms per key. Could be separate PR.

_Originally posted by @tomholub in https://github.com/FlowCrypt/flowcrypt-browser/pull/4725#discussion_r1085549366_

### PR DESCRIPTION
**Pull request by rrrooommmaaa**
This PR switches to OpenPGP v5

close #3324
close #4904
----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [x] addresses the issue it closes (if any)
- [x] code is readable and understandable
- [x] is accompanied with tests, or tests are not needed
- [x] is free of vulnerabilities
- [x] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

**Comment by tomholub** [comment]
looking great, finally :-)

are we able to drop any functionality related to dropping weak algorithms? There should be settings on v5 to do this natively.

**Comment by rrrooommmaaa** [comment]
> @rrrooommmaaa how is this looking at this moment?

I need to find a way to neatly wire the new library to the browser page, most of incompatibilities were resolved 2 months ago and unit tests are looking good.
I had to suspend the development in favor of review, now I'm able to finish it.

**Comment by rrrooommmaaa** [comment]
While waiting for https://github.com/openpgpjs/openpgpjs/pull/1583 to be merged and published, looking at other issues

**Comment by rrrooommmaaa** [comment]
@tomholub We're getting an error when trying to return `OpenPGP.Key` object from background handlers:
`Could not serialize message`
It happens because of cyclic nesting -- each of `subkeys` and `users` holds a reference to the `mainKey`.
Should we implement a workaround or... ?

**Comment by rrrooommmaaa** [comment]
> Should we implement a workaround or... ?

Perhaps, fix `OpenPGP.js` to allow transportation of the object?
For this we need to find out exactly what this `mainKey` is used for.
Perhaps, we can add a function in `OpenPGP.js` that removes all the `mainKey` references,
and add functionality in functions that search for subkeys and userids to restore this field if it's missing.

**Comment by tomholub** [comment]
I think we should not be expecting to successfully transfer OpenPGP objects across the boundary.

`OpenPGP.Message` or `OpenPGP.Key` should stay within their original boundary, and across the boundary we share the message as bytes, or key as bytes.

Which current action causes deserialized key to be sent across?

**Comment by rrrooommmaaa** [comment]
> `OpenPGP.Message` or `OpenPGP.Key` should stay within their original space. Across the boundary we only share the message as bytes, or key as bytes.
> 
> Which current action causes deserialized key to be sent across?

We have "hidden" fields in our own `Key` type: `rawKey`, `internal` and `rawArmored`.
`rawKey` is `OpenPGP.Key`, `internal` is safe to use `OpenPGP.Key` (without weak subkeys).

So wherever we pass `Key` (inside `PubkeyInfo`), we pass the underlying `OpenPGP.Key` objects.
These are returned by `ContactStore` 
- `getEncryptionKeys`
- `getOneWithAllPubkeys`
- `getPubkeyInfos`

Some `ContactStore` methods accept `Key` arguments, so it seems natural to return `Key` too.
Are you thinking about refactoring `ContactStore`?

**Comment by rrrooommmaaa** [comment]
We can also think about wiping `rawKey` and `internal` fields when sending/receiving messages,
and restoring it on-demand in `extension/js/common/core/crypto/pgp/openpgp-key.ts` functions

**Comment by tomholub** [comment]
> Some ContactStore methods accept Key arguments, so it seems natural to return Key too.
> Are you thinking about refactoring ContactStore?

I never intended to be sending OpenPGP.js objects across the wire, to my knowledge the objects undergo JSON serialization / deserialization in the process and last time I tried, that didn't work for OpenPGP.js objects. I'm surprised it worked that way.

All background methods should be refactored to not not accept or return these, and if that includes the ContactStore, than that should be refactored also to send simple objects across.

**Comment by rrrooommmaaa** [comment]
Yes, all existing tests now pass.
I wanted to thoroughly review the changes and possibly add some additional tests concerning the changed logic.

**Comment by tomholub** [comment]
Great news 👍

**Comment by tomholub** [comment]
Please also do rudimentary manual test on Firefox, to double check nothing is horribly broken, like the content script not loading or not working properly, and setup/send/receive emails with attachments.

File: extension/js/common/core/crypto/pgp/msg-util.ts [hunk 1]
```diff
@@ -145,7 +143,14 @@ export class MsgUtil {
         // This does not 100% mean it's OpenPGP message
         // But it's a good indication that it may be
         const t = opgp.enums.packet;
-        const msgTpes = [t.symEncryptedIntegrityProtected, t.modificationDetectionCode, t.symEncryptedAEADProtected, t.symmetricallyEncrypted, t.compressed];
+        const msgTpes = [
+          t.publicKeyEncryptedSessionKey,
```
Comment (rrrooommmaaa) [inline extension/js/common/core/crypto/pgp/msg-util.ts:147]: I added `publicKeyEncryptedSessionKey` type here as a pubkey-encrypted message started with this type of packet. So this method isn't used often and tested properly.
We can review and test it in a separate issue

Comment (tomholub) [reply extension/js/common/core/crypto/pgp/msg-util.ts:147]: ok

File: extension/js/common/core/crypto/key.ts [hunk 2]
```diff
@@ -221,9 +222,9 @@ export class KeyUtil {
           allErr.push(e as Error);
         }
       }
-      if (err) {
+      /* todo: re-throw? if (err) {
         allErr.push(...err);
-      }
+      } */
```
Comment (tomholub) [inline extension/js/common/core/crypto/key.ts:227]: I wanted to allow reading the results partially. Like if WKD sends a mixture of keys that some throw when processing and some don't, I'd like to see the valid keys and then the errors side by side (code-wise), so that the app may choose to still use the valid ones and ignoring the errors. That's why we are returning errors instead of throwing them, unlike everywhere else in the codebase (I suppose this is what your comment was about).

File: extension/js/common/core/crypto/pgp/openpgp-key.ts [hunk 3]
```diff
@@ -27,21 +34,32 @@ export class OpenPGPKey {
...
 
-  public static validateAllDecryptedPackets = async (key: OpenPGP.key.Key): Promise<void> => {
+  // remove crypto-library objects (useful when sending the object to/from background)
+  public static pack = (key: Key) => {
+    const keyWithPrivateFields = key as unknown as KeyWithPrivateFields;
+    if (typeof keyWithPrivateFields.internal !== 'string') {
+      keyWithPrivateFields.internal = keyWithPrivateFields.internal.armor();
+      keyWithPrivateFields.rawKey = undefined;
+    }
+  };
```
Comment (tomholub) [inline extension/js/common/core/crypto/pgp/openpgp-key.ts:58]: Simplicity looks good.

Would it be faster to send the key in binary representation (unarmored openpgp key)? Armoring/de-armoring is oddly expensive, about 10ms per key. Could be separate PR.

File: extension/js/common/core/crypto/pgp/msg-util.ts (hunk 1 above)
Comment (tomholub) [reply extension/js/common/core/crypto/pgp/msg-util.ts:147]: I don't think we necessarily have to unit-test this method, it's not like we normally edit it or plan to do so. I think it's likely broadly covered by other tests.

File: extension/js/common/core/crypto/pgp/openpgp-key.ts [hunk 4]
```diff
@@ -27,21 +34,32 @@ export class OpenPGPKey {
   };
 
   public static parseMany = async (text: string): Promise<Key[]> => {
-    const result = await opgp.key.readArmored(text);
+    const resultKeys = await opgp.readKeys({ armoredKeys: text });
+    /* todo: test exception and re-wrap?
     if (result.err) {
       throw new Error('Cannot parse OpenPGP key: ' + result.err + ' for: ' + text);
     }
```
Comment (tomholub) [inline extension/js/common/core/crypto/pgp/openpgp-key.ts:41]: Not checking this follows previous behavior, right? If so, we can file this for another issue, review usages, and decide how to handle these.

File: extension/js/common/core/crypto/pgp/openpgp-key.ts [hunk 5]
```diff
@@ -118,74 +113,67 @@ export class OpenPGPKey {
...
-    }
-    if (pwd) {
-      options.passwords = [pwd];
-    }
     if (!pubkeys && !pwd) {
       throw new Error('no-pubkeys-no-challenge');
     }
+    const message = await opgp.createMessage({ binary: data, filename, date });
+    const encryptionKeys = await Promise.all(pubkeys?.map(OpenPGPKey.extractExternalLibraryObjFromKey) ?? []);
+    // TODO: Investigate unwrapping?
```
Comment (tomholub) [inline extension/js/common/core/crypto/pgp/openpgp-key.ts:137]: Highlighting, although unsure about what the todo means

File: extension/js/common/core/crypto/pgp/openpgp-key.ts [hunk 6]
```diff
@@ -194,40 +182,25 @@ export class OpenPGPKey {
    * @param keyToUpdate - an existing Key object to update, optional. Useful in encryptKey and decryptKey, because the operation
    *    is done on the original supplied object.
    */
-  public static convertExternalLibraryObjToKey = async (opgpKey: OpenPGP.key.Key, keyToUpdate?: Key): Promise<Key> => {
+  public static convertExternalLibraryObjToKey = async (opgpKey: OpenPGP.Key, keyToUpdate?: Key): Promise<Key> => {
     const { isPrimaryKeyStrong, keyWithoutWeakPackets } = OpenPGPKey.removeWeakKeyPackets(opgpKey);
```
Comment (tomholub) [inline extension/js/common/core/crypto/pgp/openpgp-key.ts:186]: I suppose there was new functionality for this in OpenPGP.js v5, where we can simply configure which algorithms to use or forbid? Either in global config or when doing particular actions (forgot about their implementation). In which case we could stop using `removeWeakKeyPackets` / remove it. Can be another issue.

File: extension/js/common/core/crypto/pgp/openpgp-key.ts [hunk 7]
```diff
@@ -269,114 +244,99 @@ export class OpenPGPKey {
...
+    const message = await opgp.createMessage({ text });
+    const signRes = await opgp.sign({ message, format: 'armored', signingKeys: [signingPrv], detached });
+    return signRes;
   };
 
   public static revoke = async (key: Key): Promise<string | undefined> => {
-    let prv = OpenPGPKey.extractExternalLibraryObjFromKey(key);
+    let prv = await OpenPGPKey.extractExternalLibraryObjFromKey(key);
+    if (!prv.isPrivate()) {
+      return undefined; // todo: or throw?
```
Comment (tomholub) [inline extension/js/common/core/crypto/pgp/openpgp-key.ts:282]: I'd throw

File: extension/js/common/core/crypto/pgp/openpgp-key.ts [hunk 8]
```diff
@@ -269,114 +244,99 @@ export class OpenPGPKey {
...
+    }
     if (!(await prv.isRevoked())) {
-      prv = await prv.revoke({});
+      const keypair = await opgp.revokeKey({ key: prv, format: 'object' });
+      // todo: save this data into `key`?
```
Comment (tomholub) [inline extension/js/common/core/crypto/pgp/openpgp-key.ts:286]: I'm not entirely sure now what is it that we are actually doing. Did we revoke the key, or did we create a revocation certificate that can be saved and later uploaded to public key servers?

If the latter (that's what I thought), people sometimes do that as a form of forward-safety, save a revocation cert somewhere more accessible than the private key itself. not that any FlowCrypt usecases rely on that.

I think the method naming is ambiguous, UI says `revocation certificate` but this method says `revoke` when it should say `newRevocationCertificate` or similar, in which case your note would not apply.

File: extension/js/common/core/crypto/pgp/openpgp-key.ts [hunk 9]
```diff
@@ -505,40 +498,28 @@ export class OpenPGPKey {
...
+    const opgpKeys = await Promise.all(validKeys.map(x => OpenPGPKey.extractExternalLibraryObjFromKey(x)));
     // todo: expired?
     try {
-      // this is here to ensure execution order when 1) verify, 2) read data, 3) processing signatures
-      // Else it will hang trying to read a stream: https://github.com/openpgpjs/openpgpjs/issues/916#issuecomment-510620625
-      const verifications = await msg.verify(opgpKeys); // first step
-      const stream = msg instanceof opgp.message.Message ? msg.getLiteralData() : msg.getText();
-      if (stream) {
+      const signerLongids = msg.getSigningKeyIDs().map(kid => OpenPGPKey.bytesToLongid(kid.bytes));
+      const text = msg instanceof opgp.CleartextMessage ? msg.getText() : msg.getLiteralData(); // todo: is this important?
```
Comment (tomholub) [inline extension/js/common/core/crypto/pgp/openpgp-key.ts:507]: On v4 it was, at least. `getText` was for cleartext signed data, `getLiteralData` was for signed encrypted data :shrug:

File: extension/js/common/core/crypto/pgp/openpgp-key.ts [hunk 10]
```diff
@@ -618,127 +601,154 @@ export class OpenPGPKey {
...
+      return keyWithPrivateFields.internal;
     }
-    return opgpKey;
+    return internal;
   };
 
-  private static extractStrengthUncheckedExternalLibraryObjFromKey = (pubkey: Key) => {
-    if (pubkey.family !== 'openpgp') {
-      throw new UnexpectedKeyTypeError(`Key type is ${pubkey.family}, expecting OpenPGP`);
+  private static extractStrengthUncheckedExternalLibraryObjFromKey = async (key: Key) => {
```
Comment (tomholub) [inline extension/js/common/core/crypto/pgp/openpgp-key.ts:724]: Also these "StrengthUnchecked" etc checks could be removed if the above comment is true. Doesn't have to be now.

File: extension/js/content_scripts/webmail/gmail-element-replacer.ts [hunk 11]
```diff
@@ -16,12 +16,14 @@ import { Gmail } from '../../common/api/email-provider/gmail/gmail.js';
 import { Injector } from '../../common/inject.js';
 import { PubLookup } from '../../common/api/pub-lookup.js';
 import { Notifications } from '../../common/notifications.js';
+// note: only types are supported for OpenPGP.js or web-stream-tools, their function calls will fail
```
Comment (tomholub) [inline extension/js/content_scripts/webmail/gmail-element-replacer.ts:19]: This is because last time we tried to use OpenPGP.js in the content script, it didn't work (on Firefox). Actually, it would be nice to load it there. Then we wouldn't have to make so many calls to background_script (which we have to eventually deprecate anyway because of Google Chrome extension manifest v3. https://github.com/FlowCrypt/flowcrypt-browser/issues/1509)

I'll file an issue to try to attempt running openpgp.js in content script again, at which point this comment could be removed, and some things simiplified, in particular:
 - background calls from content script to background page that are only done because OpenPGP.js can be removed, and instead appropriate methods (which rely on OpenPGP.js) can be called directly
 - pgp_block.htm/ts that today calls into background page can instead call the content script when it needs OpenPGP.js, and those background call handlers can be moved from background page to content script

File: extension/js/content_scripts/webmail/gmail-element-replacer.ts [hunk 12]
```diff
@@ -16,12 +16,14 @@ import { Gmail } from '../../common/api/email-provider/gmail/gmail.js';
...
 import { PgpArmor } from '../../common/core/crypto/pgp/pgp-armor.js';
 import { Ui } from '../../common/browser/ui.js';
 import { WebmailCommon } from '../../common/webmail.js';
 import { Xss } from '../../common/platform/xss.js';
 import { ClientConfiguration } from '../../common/client-configuration.js';
 import { SendAsAlias } from '../../common/platform/store/acct-store.js';
+// todo: can we somehow define a purely relay class for ContactStore to clearly show that crypto-libraries are not loaded and can't be used?
 import { ContactStore } from '../../common/platform/store/contact-store.js';
```
Comment (tomholub) [inline extension/js/content_scripts/webmail/gmail-element-replacer.ts:27]: Good idea, we could separate it, which would be less messy. I'm not filing an issue yet, we'll leave this comment here, and see what to do in the context of https://github.com/FlowCrypt/flowcrypt-browser/issues/1509

**Review by tomholub — COMMENTED** [review]
Looking great, with some comments, mostly for the future

File: extension/js/common/core/crypto/pgp/msg-util.ts [hunk 13]
```diff
@@ -145,7 +143,14 @@ export class MsgUtil {
         // This does not 100% mean it's OpenPGP message
         // But it's a good indication that it may be
         const t = opgp.enums.packet;
-        const msgTpes = [t.symEncryptedIntegrityProtected, t.modificationDetectionCode, t.symEncryptedAEADProtected, t.symmetricallyEncrypted, t.compressed];
+        const msgTpes = [
```
Comment (sosnovsky) [inline extension/js/common/core/crypto/pgp/msg-util.ts:146]: It's probably should be `msgTypes`?

File: extension/js/common/core/crypto/pgp/msg-util.ts [hunk 14]
```diff
@@ -213,10 +218,12 @@ export class MsgUtil {
...
       }
       // cleartext and PKCS#7 are gone by this line
-      const msg = prepared.message as OpenPGP.message.Message;
+      const msg = prepared.message as OpenPGP.Message<OpenPGP.Data>;
       const packets = msg.packets;
-      const isSymEncrypted = packets.filter(p => p.tag === opgp.enums.packet.symEncryptedSessionKey).length > 0;
-      const isPubEncrypted = packets.filter(p => p.tag === opgp.enums.packet.publicKeyEncryptedSessionKey).length > 0;
+      // todo: remove this patch after openpgpjs/openpgpjs#1583 is resolved
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      const isSymEncrypted = packets.filter(p => p instanceof (opgp as any).SymEncryptedSessionKeyPacket).length > 0; // tslint:disable-line:no-unsafe-any
```
Comment (sosnovsky) [inline extension/js/common/core/crypto/pgp/msg-util.ts:225]: `tslint` comments not needed anymore

File: package.json [hunk 15]
```diff
@@ -34,14 +35,16 @@
...
     "pdfjs-dist": "3.2.146",
     "prettier": "^2.8.3",
     "puppeteer": "19.5.2",
+    "stream-browserify": "^3.0.0",
     "stylelint": "14.16.1",
     "stylelint-config-standard": "29.0.0",
     "typescript": "4.9.4",
-    "web-ext": "7.4.0"
+    "web-ext": "7.4.0",
+    "webpack-cli": "^5.0.1"
```
Comment (sosnovsky) [inline package.json:47]: Is `webpack` used somewhere in the project? Haven't found usage in the code

File: extension/js/common/core/crypto/pgp/openpgp-key.ts [hunk 16]
```diff
@@ -269,114 +244,99 @@ export class OpenPGPKey {
...
-            await key.decrypt(passphrase); // throws on password mismatch
-            return 'success';
-          } catch (e) {
-            if (e instanceof Error && e.message.toLowerCase().includes('incorrect key passphrase')) {
-              return 'INCORRECT PASSPHRASE';
-            } else {
-              throw e;
-            }
-          }
+          return (await OpenPGPKey.decryptPrivateKey(key, passphrase)) ? 'success' : 'INCORRECT PASSPHRASE';
```
Comment (sosnovsky) [inline extension/js/common/core/crypto/pgp/openpgp-key.ts:335]: Are other error types ignored intentionally here? As previously errors were rethrown

File: extension/js/common/core/crypto/pgp/openpgp-key.ts [hunk 17]
```diff
@@ -238,21 +211,23 @@ export class OpenPGPKey {
     if (!fingerprint) {
       throw new Error('Key does not have a fingerprint and cannot be parsed.');
     }
-    const algoInfo = keyWithoutWeakPackets.primaryKey.getAlgorithmInfo();
-    const key = keyToUpdate || ({} as Key); // if no key to update, use empty object, will get props assigned below
-
+    const algoInfo = keyWithoutWeakPackets.keyPacket.getAlgorithmInfo();
+    // tslint:disable-next-line:no-unnecessary-initializer
```
Comment (sosnovsky) [inline extension/js/common/core/crypto/pgp/openpgp-key.ts:215]: `tslint:disable` not needed

**Review by sosnovsky — COMMENTED** [review]
Wow, it's a massive PR, looks great!

File: extension/js/background_page/background_page.ts [hunk 18]
```diff
@@ -20,8 +19,6 @@ import { migrateGlobal, moveContactsToEmailsAndPubkeys, updateOpgpRevocations, u
 
 console.info('background_process.js starting');
 
-opgp.initWorker({ path: '/lib/openpgp.worker.js' });
-
```
[the remark this answers is not in the document]
Comment (tomholub) [reply extension/js/background_page/background_page.ts:24]: https://github.com/FlowCrypt/flowcrypt-browser/issues/4903

File: extension/js/common/core/crypto/pgp/openpgp-key.ts (hunk 3 above)
Comment (tomholub) [reply extension/js/common/core/crypto/pgp/openpgp-key.ts:58]: https://github.com/FlowCrypt/flowcrypt-browser/issues/4904

File: extension/js/common/core/crypto/pgp/openpgp-key.ts (hunk 6 above)
Comment (tomholub) [reply extension/js/common/core/crypto/pgp/openpgp-key.ts:186]: https://github.com/FlowCrypt/flowcrypt-browser/issues/4905

File: extension/js/common/core/crypto/pgp/openpgp-key.ts (hunk 10 above)
Comment (tomholub) [reply extension/js/common/core/crypto/pgp/openpgp-key.ts:724]: #4905

File: extension/js/content_scripts/webmail/gmail-element-replacer.ts (hunk 11 above)
Comment (tomholub) [reply extension/js/content_scripts/webmail/gmail-element-replacer.ts:19]: https://github.com/FlowCrypt/flowcrypt-browser/issues/4906

File: test/source/tests/unit-node.ts [hunk 19]
```diff
@@ -1291,19 +1353,18 @@ ZAvn6PBX7vsaReOVa2zsnuY5g70xCxvzHIwR94POu5cENwRtCkrppFnISALpQ1kA
       expect(result.get('Sign/Verify test')).to.equal('[-] skipped, not fully decrypted');
       expect(result.get('SK 0 > LongId')).to.equal('[-] 0485D618EAA64B05');
       expect(result.get('SK 0 > Created')).to.equal('[-] 1606140328 or 2020-11-23T14:05:28.000Z');
-      expect(result.get('SK 0 > Algo')).to.equal('[-] rsa_encrypt_sign');
+      expect(result.get('SK 0 > Algo')).to.equal('[-] rsaEncryptSign');
       expect(result.get('SK 0 > Verify')).to.equal('[-] OK');
-      expect(result.get('SK 0 > Subkey tag')).to.equal('[-] 7');
+      expect(result.get('SK 0 > Subkey object type')).to.equal('[-] SecretSubkeyPacket');
```
Comment (rrrooommmaaa) [inline test/source/tests/unit-node.ts:1358]: There might be a way to restore the original behavior using `constructor::tag`.
I haven't tried it yet, should I, @tomholub ?

File: package.json (hunk 15 above)
Comment (rrrooommmaaa) [reply package.json]: I removed the extra packages

File: extension/js/common/core/crypto/pgp/msg-util.ts (hunk 14 above)
Comment (rrrooommmaaa) [reply extension/js/common/core/crypto/pgp/msg-util.ts]: removed it

File: extension/js/common/core/crypto/pgp/msg-util.ts (hunk 13 above)
Comment (rrrooommmaaa) [reply extension/js/common/core/crypto/pgp/msg-util.ts]: renamed it

File: test/source/tests/unit-node.ts (hunk 19 above)
Comment (tomholub) [reply test/source/tests/unit-node.ts:1358]: this looks good as is

File: extension/js/common/core/crypto/pgp/openpgp-key.ts (hunk 4 above)
Comment (rrrooommmaaa) [reply extension/js/common/core/crypto/pgp/openpgp-key.ts:41]: Yes, this generally follow the previous behavior, but less descriptive, e.g.
`Misformed armored text` instead of `Cannot parse OpenPGP key: Misformed armored text for: ...`
I added unit tests for this.

File: build.sh [hunk 20]
```diff
@@ -65,14 +66,20 @@ mkdir -p $OUTDIR/lib/bootstrap
 cp node_modules/bootstrap/dist/js/bootstrap.min.js $OUTDIR/lib/bootstrap/bootstrap.min.js
 cp node_modules/bootstrap/dist/css/bootstrap.min.css $OUTDIR/lib/bootstrap/bootstrap.min.css
 
+mkdir -p $OUTDIR/lib/streams
+cp node_modules/@openpgp/web-stream-tools/lib/*.js $OUTDIR/lib/streams
+# patch imports with .js, e.g. replace './streams' with './streams.js'
+# until https://github.com/openpgpjs/web-stream-tools/pull/20 is resolved
+sed -i -E "s/'\.\/(streams|util|writer|reader|node-conversions)'/'\.\/\1\.js'/g" $OUTDIR/lib/streams/*
```
Comment (rrrooommmaaa) [inline build.sh:73]: I can add a more complex regex to search for
`import { ... } from './file'`
and `import * as ... from './file'`

What do you think?

File: extension/js/common/core/crypto/pgp/openpgp-key.ts (hunk 8 above)
Comment (rrrooommmaaa) [reply extension/js/common/core/crypto/pgp/openpgp-key.ts]: Right, I renamed this method to `getOrCreateRevocationCertificate()` as it either extracts the revocation certificate from the key (both public and private) if it's already revoked, or creates a new revocation certificate (if the key is private, otherwise throws an exception).
I also added `applyRevocationCertificate()` method and a unit test.

File: extension/js/common/core/crypto/pgp/openpgp-key.ts (hunk 3 above)
Comment (rrrooommmaaa) [reply extension/js/common/core/crypto/pgp/openpgp-key.ts]: I added this into this PR

File: extension/js/common/core/crypto/pgp/openpgp-key.ts (hunk 7 above)
Comment (rrrooommmaaa) [reply extension/js/common/core/crypto/pgp/openpgp-key.ts]: Done

File: extension/js/common/core/crypto/key.ts [hunk 21]
```diff
@@ -203,7 +204,7 @@ export class KeyUtil {
       allErr: Error[] = [];
     let uncheckedOpgpKeyCount = 0;
     try {
-      const { keys, err } = await opgp.key.read(key);
+      const keys = await opgp.readKeys({ binaryKeys: key }); // todo: opgp.readKey ?
```
[the remark this answers is not in the document]
Comment (rrrooommmaaa) [reply extension/js/common/core/crypto/key.ts:207]: The new `opgp.readKeys` function differs in that it doesn't wrap each `new PrivateKey/new PublicKey` constructor in a `try/catch` block unlike the 4.10.10 implementation. So if we stumble on the some non-last key, the remaining keys won't be processed.
That got me confused

File: extension/js/common/core/crypto/key.ts (hunk 2 above)
Comment (rrrooommmaaa) [reply extension/js/common/core/crypto/key.ts:227]: The results may differ compared to 4.10.10 implementation so we need to decide how to deal with this (see comment above).

File: extension/js/common/core/crypto/pgp/openpgp-key.ts (hunk 4 above)
Comment (rrrooommmaaa) [reply extension/js/common/core/crypto/pgp/openpgp-key.ts:41]: And also a concern about premature processing stop (same as with `KeyUtil.readBinary`)

File: extension/js/common/core/crypto/pgp/openpgp-key.ts (hunk 16 above)
Comment (rrrooommmaaa) [reply extension/js/common/core/crypto/pgp/openpgp-key.ts:341]: If an exception occurs, `formatResultAsync` will display the exception data in square brackets
```
    } catch (e) {
      return `[${String(e)}]`;
    }
```

File: build.sh (hunk 20 above)
Comment (tomholub) [reply build.sh:73]: It's a temporary fix, right? I think ok as is.

**Review by tomholub — DISMISSED** [review]
reviewing delta

File: extension/js/common/core/crypto/key.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (tomholub) [reply extension/js/common/core/crypto/key.ts:207]: Ah. Not ideal but ok to proceed as is and file an issue.

File: extension/js/common/core/crypto/key.ts (hunk 2 above)
Comment (tomholub) [reply extension/js/common/core/crypto/key.ts:227]: let's file the comment above for another issue.

File: extension/js/common/core/crypto/key.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (tomholub) [reply extension/js/common/core/crypto/key.ts:207]: also https://github.com/FlowCrypt/flowcrypt-browser/pull/4725#discussion_r1089731626

File: extension/js/common/core/crypto/pgp/openpgp-key.ts (hunk 4 above)
Comment (tomholub) [reply extension/js/common/core/crypto/pgp/openpgp-key.ts:41]: mentioned here https://github.com/FlowCrypt/flowcrypt-browser/pull/4725#discussion_r1090981268

File: extension/js/common/core/crypto/pgp/openpgp-key.ts (hunk 8 above)
Comment (tomholub) [reply extension/js/common/core/crypto/pgp/openpgp-key.ts]: sounds good

File: extension/js/common/core/crypto/pgp/openpgp-key.ts [hunk 22]
```diff
@@ -618,127 +610,150 @@ export class OpenPGPKey {
...
-      opgp.enums.signature.subkey_binding,
-      dataToVerify,
-      date
-    );
+    const bindingSignature = await OpenPGPKey.getLatestValidSignature(subKey.bindingSignatures, primaryKey, opgp.enums.signature.subkeyBinding, dataToVerify);
     if (
       bindingSignature &&
       bindingSignature.embeddedSignature &&
-      (await OpenPGPKey.getLatestValidSignature([bindingSignature.embeddedSignature], subKey.keyPacket, opgp.enums.signature.key_binding, dataToVerify, date))
+      (await OpenPGPKey.getLatestValidSignature([bindingSignature.embeddedSignature], subKey.keyPacket, opgp.enums.signature.keyBinding, dataToVerify))
```
[the remark this answers is not in the document]
Comment (tomholub) [reply extension/js/common/core/crypto/pgp/openpgp-key.ts:655]: I'm not sure and I'm ok to not worry about it if it passes existing tests. However, it would be good to file it for another issue to consider further.

File: extension/js/common/core/crypto/pgp/openpgp-key.ts [hunk 23]
```diff
@@ -618,178 +607,226 @@ export class OpenPGPKey {
...
-    if (firstTry) {
-      return firstTry;
+  private static getSigningAndEncryptionKeys = async (key: OpenPGP.Key) => {
+    // encryptionKey and signingKey are obtained with undefined value of the date parameter, which is considered current date
+    const encryptionKey = await Catch.undefinedOnException(key.getEncryptionKey());
+    const signingKey = await Catch.undefinedOnException(key.getSigningKey());
+    const possibleExpirations: number[] = [];
+    const primaryKeyExpiration = OpenPGPKey.getExpirationAsDateOrUndefined(await key.getExpirationTime())?.getTime();
+    if (!encryptionKey || !signingKey) {
+      possibleExpirations.push(
```
Comment (rrrooommmaaa) [inline extension/js/common/core/crypto/pgp/openpgp-key.ts:744]: I reworked this method to make use of `getSigningKey` and `getEncryptionKey` calls from OpenPGP.js instead of trying to replicate OpenPGP.js behaviour when selecting keys, but it turned out to be not so simple. Supplying `date: null` (meaning: ignore date) parameter instead of `date: undefined` (meaning: take current date) it is possible to find a (sub)key for singing/encryption, but it may not be the subkey we're looking for ( 1) the subkey may have `created` property after primary key's expiration is in the `key was never usable` unit test, 2) the subkey may tell us an incorrect expiration date, as there may be another subkey with later expiration that didn't show up because OpenPGP.js sorts subkeys by `created` property)
So, for already expired keys the following algo is implemented:
1. create a list of all the subkey's expiration dates (prior to primary key's expiration) 
(todo: we can make it faster by manually collecting expirations from signatures?)
2. call  `getEncryptionKey`/`getSigningKey`  with dates from the list in descending order until we get a usable key.
If a not-expired encryption key is returned for the current date, then we'll call `getEncryptionKey` with the expiration date of the found key to find a next valid key to figure out the final `expiration` date over all the subkeys.
What do you think, @tomholub ? Is this approach good?

**Review by tomholub — COMMENTED** [review]
delta review - looks good

File: extension/js/common/core/crypto/pgp/openpgp-key.ts (hunk 5 above)
Comment (rrrooommmaaa) [reply extension/js/common/core/crypto/pgp/openpgp-key.ts:137]: Perhaps, it's outdated.
The keys coming from the background are now `wrapped` (packed) as of this PR

File: extension/js/common/core/crypto/pgp/openpgp-key.ts (hunk 23 above)
Comment (rrrooommmaaa) [reply extension/js/common/core/crypto/pgp/openpgp-key.ts:759]: Done

**Comment by rrrooommmaaa** [comment]
> Please also do rudimentary manual test on Firefox, to double check nothing is horribly broken, like the content script not loading or not working properly, and setup/send/receive emails with attachments.

I tested the extension in Firefox and it seems to setup and send/receive messages with attachments ok

**Review by tomholub — APPROVED** [review]
